bytestring_splitter/__init__.py

```diff
@@ -379,6 +378,210 @@ def _parse_message_meta(message_item):
+        """
+        for subcls in cls._get_ordered_mixins():
+            some_bytes = subcls._strip_metadata(some_bytes)
+        return some_bytes
+
+    @classmethod
+    def _assign_metadata(cls, some_bytes, **kwargs):
+        """
+        called by the baseclass for all subclasses for them to add their own metadata
+        """
+        data = kwargs.get(cls.METADATA_TAG, None) or\
+            getattr(cls, f'_input_{cls.METADATA_TAG}', None) or\
+            getattr(cls, cls.METADATA_TAG, None)
+
+        if not data:
+            raise ValueError(f"could not determine {cls.METADATA_TAG} to assign to output bytes and none was supplied")
+        return cls._prepend_metadata(cls._serialize_metadata(data), some_bytes)
+
+    @classmethod
+    def _prepend_metadata(cls, data, some_bytes):
+        return data + some_bytes
+
+    @classmethod
+    def _strip_metadata(self, some_bytes):
+        return some_bytes[self.HEADER_LENGTH:]
+
+    @classmethod
+    def _get_metadata(cls, some_bytes, data=None):
+        # gets the metadata off the top of the bytestring for this mixin
+        data = data or {}
+        data_bytes = some_bytes[:cls.HEADER_LENGTH]
+        data[cls.METADATA_TAG] = cls._deserialize_metadata(data_bytes)
+
+        return data
+
+    def get_header_bytes(self, some_bytes):
+        return self._get_metadata(some_bytes)[self.METADATA_TAG]
+
+    @classmethod
+    def _deserialize_metadata(cls, data_bytes):
+        """
+        will often be overridden to transform metadata as needed
+        for a given type of metadata
+
+        see VersioningMixin below
+        """
+
+        return data_bytes
+
+    @classmethod
+    def _serialize_metadata(cls, data_bytes):
+        """
+        will often be overridden to transform metadata as needed
+        for a given type of metadata
+
+        see VersioningMixin below
+        """
+        return data_bytes
+
+
+class VersioningMixin(HeaderMetaDataMixinBase):
+
+    HEADER_LENGTH = 2
+    METADATA_TAG = 'version'
+
+    @classmethod
+    def _deserialize_metadata(cls, data_bytes):
+        return int.from_bytes(data_bytes, 'big')
+
+    @classmethod
+    def _serialize_metadata(cls, value):
+        return value.to_bytes(cls.HEADER_LENGTH, "big")
+
+    @classmethod
+    def assign_version(cls, some_bytes, version):
+        #  a convenience method specific to VersioningMixin
+        return cls.assign_metadata(some_bytes, version=version)
+
+
+class StructureChecksumMixin(HeaderMetaDataMixinBase):
+
+    HEADER_LENGTH = 4
+    METADATA_TAG = 'checksum'
+
+    class InvalidBytestringException(BaseException):
+        pass
+
+    def generate_checksum(self, *args, **kwargs):
+        hash = hashlib.sha256()
+        for mt in self.message_types:
+            message_name, message_class, message_length, kwargs = mt
+            hash.update(
+                b'v' if message_length is VariableLengthBytestring
```

Comment:
```suggestion
                b'\xFF\xFF\xFF\xFF' if message_length is VariableLengthBytestring
```
I'd change this to a 4-byte long bytestring, as the other inputs to this hash function (which has no negative effect whatsoever), to disallow trivial hash collisions such as:
```python
def test_checksum_collision():

    four_variables_splitter = ChecksumVerifyingSplitter(
        VariableLengthBytestring,
        VariableLengthBytestring,
        VariableLengthBytestring,
        VariableLengthBytestring,
    )

    hd_movie_splitter = ChecksumVerifyingSplitter(
        (bytes, 1987475062)
    )

    assert four_variables_splitter.generate_checksum() != hd_movie_splitter.generate_checksum()
```
This test currently fails.

Reply:
@cygnusv  good idea!

Reply:
Hmm... could probably figure out a way to salt those...